### PR TITLE
ADBDEV-5489: Make InitPlans non-rescannable (GPDB 7)

### PR DESCRIPTION
**Pull request by bandetto**
This is a cherry-pick of f42c552. This PR should be rebased to preserve authorship.

The original commit was adapted for GPDB 7. All ORCA plans are now marked with REWIND flag, since it can't generate InitPlans at all (refer to e7cc27f). Changes in the ORCA code were modified to allow extracting subplans inside subplans, since extract_nodes_plan() can't do it. The test was simplified.

### CODE REVIEW

File: src/backend/executor/execMain.c [hunk 1]
```diff
@@ -1998,12 +1998,20 @@ InitPlan(QueryDesc *queryDesc, int eflags)
...
 			sp_eflags = eflags
 				& (EXEC_FLAG_EXPLAIN_ONLY | EXEC_FLAG_WITH_NO_DATA);
-			sp_eflags |= EXEC_FLAG_REWIND;
+			
+			/*
+			 * GPDB: For ORCA, we always set the REWIND flag to delay
+			 * eagerfree. However, planner can generate InitPlans, which are
+			 * non-rewindable.
+			 */
+			if (estate->es_plannedstmt->planGen == PLANGEN_OPTIMIZER ||
```
Comment (Stolb27) [inline src/backend/executor/execMain.c]: I think it's a weak solution in the context of https://github.com/arenadata/gpdb/pull/1124/files#diff-2658a1ae6156cefca01f7e0fe169073446b923caa722c61dbf33503854860458

File: src/backend/gpopt/translate/CTranslatorDXLToPlStmt.cpp [hunk 2]
```diff
@@ -180,13 +180,13 @@ CTranslatorDXLToPlStmt::GetPlannedStmtFromDXL(const CDXLNode *dxlnode,
 	// collect unique RTE in FROM Clause
 	List *oids_list_unique = NIL;
 
-	ListCell *lc_rte = nullptr;
+	ListCell *lc = nullptr;
```
Comment (whitehawk) [inline src/backend/gpopt/translate/CTranslatorDXLToPlStmt.cpp:183]: This renaming looks irrelevant to the patch. It is not a part of original commit. Why do we need it?

File: src/backend/executor/execMain.c [hunk 3]
```diff
@@ -1998,12 +1998,17 @@ InitPlan(QueryDesc *queryDesc, int eflags)
...
 			sp_eflags = eflags
 				& (EXEC_FLAG_EXPLAIN_ONLY | EXEC_FLAG_WITH_NO_DATA);
-			sp_eflags |= EXEC_FLAG_REWIND;
+
+			/*
+			 * GPDB: For ORCA, we always set the REWIND flag to delay
+			 * eagerfree. However, planner can generate InitPlans, which are
+			 * non-rewindable.
+			 */
+			if (bms_is_member(subplan_id, plannedstmt->rewindPlanIDs))
```
Comment (whitehawk) [inline src/backend/executor/execMain.c:2010]: In the original patch `subplan_idx + 1` is used. Why is it changed? Shouldn't it be reflected in the commit description (in the section with changes comparing to the original commit)?

**Comment by whitehawk** [comment]
> All ORCA plans are now marked with REWIND flag, since it can't generate InitPlans at all.

Where is it stated? Can you please provide a reference to code/documentation?

File: src/backend/executor/execMain.c (hunk 3 above)
Comment (bandetto) [reply src/backend/executor/execMain.c:2010]: In GPDB 7, the `subplan_id` variable is assigned to 1 on initialization, and in some places it is used as `subplan_id - 1`. In GPDB 6, it is assigned to 0 before the usage and is used as is in places I pointed out before.

**Comment by bandetto** [comment]
> Where is it stated? Can you please provide a reference to code/documentation?

For example, the commit e7cc27f. I updated the PR and commit's description.

**Review by whitehawk — DISMISSED** [review]

File: src/backend/gpopt/translate/CTranslatorDXLToPlStmt.cpp [hunk 4]
```diff
@@ -245,6 +245,18 @@ CTranslatorDXLToPlStmt::GetPlannedStmtFromDXL(const CDXLNode *dxlnode,
 
 	topslice = &planned_stmt->slices[0];
 
+	// Always set the REWIND flag for subplans, since ORCA cannot generate
+	// InitPlans.
+	List *subplan_list =
+		gpdb::ExtractNodesPlan(planned_stmt->planTree, T_SubPlan, true);
```
Comment (bimboterminator1) [inline src/backend/gpopt/translate/CTranslatorDXLToPlStmt.cpp]: I noticed that I missed one unclear aspect of ExtractNodesPlan in original patch. The case is that `extract_nodes_plan` [does not walk into SubPlan subquery](https://github.com/arenadata/gpdb/blob/e8aa067ffdaa4618d8483997ecf341c6d1ea0969/src/backend/optimizer/util/walkers.c#L701-L707), and it leads to incomplete subplan processing in original and current patches, when one SubPlan contains another one inside. Like

<details>
  <summary>Ex</summary>

  ```

create table t1 (i int, j int);
create table t2 (i int, j int);
create table t3 (i int, j int) distributed randomly;
explain select * from t1 where t1.i > ALL (select t2.j from t2 where t2.i=t1.i and t2.j > ALL (select i from t3 where t3.i=t2.i));
Result  (cost=0.00..1356691736.76 rows=1 width=8)
   Filter: (SubPlan 2)
   ->  Gather Motion 3:1  (slice1; segments: 3)  (cost=0.00..431.00 rows=1 width=8)
         ->  Seq Scan on t1  (cost=0.00..431.00 rows=1 width=8)
   SubPlan 2
     ->  Result  (cost=0.00..1324032.27 rows=1 width=1)
           Filter: ((CASE WHEN (sum((CASE WHEN (t1.i <= t2.j) THEN 1 ELSE 0 END)) IS NULL) THEN true WHEN (sum((CASE WHEN (t2.j IS NULL) THEN 1 ELSE 0 END)) > '0'::bigint) THEN NULL::boolean
 WHEN (t1.i IS NULL) THEN NULL::boolean WHEN (sum((CASE WHEN (t1.i <= t2.j) THEN 1 ELSE 0 END)) = '0'::bigint) THEN true ELSE false END) = true)
           ->  Aggregate  (cost=0.00..1324032.27 rows=1 width=16)
                 ->  Result  (cost=0.00..1324032.27 rows=1 width=4)
                       Filter: (t2.i = t1.i)
                       ->  Materialize  (cost=0.00..1324032.27 rows=1 width=8)
                             ->  Gather Motion 3:1  (slice2; segments: 3)  (cost=0.00..1324032.27 rows=1 width=8)
                                   ->  Seq Scan on t2  (cost=0.00..1324032.27 rows=1 width=8)
                                         Filter: (SubPlan 1)
                                         SubPlan 1
                                           ->  Result  (cost=0.00..431.00 rows=1 width=1)
                                                 Filter: ((CASE WHEN (sum((CASE WHEN (t2.j <= t3.i) THEN 1 ELSE 0 END)) IS NULL) THEN true WHEN (sum((CASE WHEN (t3.i IS NULL) THEN 1 ELSE 0 E
ND)) > '0'::bigint) THEN NULL::boolean WHEN (t2.j IS NULL) THEN NULL::boolean WHEN (sum((CASE WHEN (t2.j <= t3.i) THEN 1 ELSE 0 END)) = '0'::bigint) THEN true ELSE false END) = true)
                                                 ->  Aggregate  (cost=0.00..431.00 rows=1 width=16)
                                                       ->  Result  (cost=0.00..431.00 rows=1 width=4)
                                                             Filter: (t3.i = t2.i)
                                                             ->  Materialize  (cost=0.00..431.00 rows=1 width=4)
                                                                   ->  Broadcast Motion 3:3  (slice3; segments: 3)  (cost=0.00..431.00 rows=1 width=4)
                                                                         ->  Seq Scan on t3  (cost=0.00..431.00 rows=1 width=4)
 Optimizer: GPORCA
(24 rows)

```
</details>

Therefore, exectutor does not initialize nested subplan with rewind flag. It's still a little bit unclear how critial is this but for correctness we should fix it.

Probably we should rework approach on ORCA  side (the planner works ok) here and create ticket for backporting it to 6.x. 

I suggest variant with storing subplan ids inside the `m_dxl_to_plstmt_context` "on fly", during creation of each SubPlan node in TranslateSubplanFromChildPlan.

**Review by bimboterminator1 — DISMISSED** [review]

File: src/backend/optimizer/plan/subselect.c [hunk 5]
```diff
@@ -785,7 +785,7 @@ build_subplan(PlannerInfo *root, Plan *plan, PlannerInfo *subroot,
 	 * there's no point since it won't get re-run without parameter changes
 	 * anyway.  The input of a hashed subplan doesn't need REWIND either.
 	 */
-	if (splan->parParam == NIL && !splan->is_initplan && !splan->useHashTable)
+	if (!splan->is_initplan)
```
Comment (whitehawk) [inline src/backend/optimizer/plan/subselect.c:785]: I think the comment above is better to update, so it is aligned with the `if` condition.

**Comment by whitehawk** [comment]
Why the commit summary was updated comparing to the original?
Please add to the commit summary a reference to this PR.

**Comment by bandetto** [comment]
> Why the commit summary was updated comparing to the original? Please add to the commit summary a reference to this PR.

I don't think I understood you well. What's a commit summary? How would you like the reference to be added?

**Comment by whitehawk** [comment]
> > Why the commit summary was updated comparing to the original? Please add to the commit summary a reference to this PR.
> 
> I don't think I understood you well. What's a commit summary? How would you like the reference to be added?

You are going to rebase the [commit](https://github.com/arenadata/gpdb/pull/1125/commits/4b98e689ac1becd4c746a1604100f95e47eec993). The summary of the commit is "Make InitPlans non-rescannable". 
The original [commit](https://github.com/arenadata/gpdb/commit/f42c552c2e9cddba8679bd5ae908ea1fba4e44f2) summary is "Make InitPlans non-rescannable during estate initialization (#595)".

I guess your commit summary should be changed from "Make InitPlans non-rescannable" to "Make InitPlans non-rescannable during estate initialization (#1125)".

**Comment by bandetto** [comment]
> I guess your commit summary should be changed from "Make InitPlans non-rescannable" to "Make InitPlans non-rescannable during estate initialization (#1125)".

Thanks for clarification. I brought back the original commit message.

**Review by whitehawk — APPROVED** [review]

**Review by bimboterminator1 — APPROVED** [review]
